matchmaking/matchmaker.py

```diff
@@ -0,0 +1,344 @@
+
+        self._match_requests_profile_to_topic = match_requests_profile_to_topic
+        # example: {
+        #   'sydneynoh@gmail.com': {'javascript', 'html-css', 'python', 'devops'}, ...
+
+        self._match_requests_topic_to_profile = match_requests_topic_to_profile
+        # example: {
+        #   'devops': {'ant@gmail.com', 'colin@gmail.me', 'sydneynoh@gmail.com'}, ...
+
+        self._possible_matches = possible_matches
+        # example: {
+        #   'sydneynoh@gmail.com': {'ant@gmail.com', 'colin@gmail.me'}, ...
+
+    def _pair_score_and_topic(self, p1, p2):
+        p1p2_key = key(p1, p2)
+        if p1p2_key in self._memos_for_pair_score_and_topic:
+            memo = self._memos_for_pair_score_and_topic[p1p2_key]
+        else:
+            memo = self._memos_for_pair_score_and_topic[p1p2_key] = self._not_memoized_pair_score_and_topic(p1, p2)
+
+        return memo
+
+    def _not_memoized_pair_score_and_topic(self, p1, p2):
+        """returns score and topic for this player to player match"""
+        assert p1 is not p2, "must not match a person with themselves"
+
+        score = 1.0  # base score
+
+        p1_topics = self._match_requests_profile_to_topic[p1]
+        p2_topics = self._match_requests_profile_to_topic[p2]
+        topics = p1_topics.intersection(p2_topics)
+
+        # if this fails then we are being very inefficient about picking pairs to test
+        assert len(topics) > 0, "won't score pair with no common topics"
+
+        # promote matching of people who have never participated before or haven't participated recently
+        score += self._boost_based_upon_recency_of_match(p1, p2)
+        # demote matching between two people that have been paired recently
+        penalty = self._penalty_based_on_recent_pairing(p1, p2)
+        score -= penalty
+        if score <= 0:
+            return 0, None
+        met_recently = False
+        if penalty:
+            met_recently = True
+
+        # get best topic
+        topic = self._select_topic(p1, p2, met_recently, topics)
+
+        return score, topic
+
+    def _boost_based_upon_recency_of_match(self, profile1, profile2):
+        # for individual players, if they've never participated, then get a boost of 1.0
+        # if they've participated recently then get a boost of 0
+        boost = 0
+        for profile in [profile1, profile2]:
+            if key(profile) in self._most_recent_match_by_profile:
+                time_since_last_meeting = self._now - self._most_recent_match_by_profile[key(profile)]['date']
+                # if time_since_last_meeting == 0
+                #   then the boost is 0
+                # if time_since_last_meeting == self._recent_match_boost_halflife
+                #   then the penalty is 0.5
+                # if time_since_last_meeting == 2*_recent_match_boost_halflife
+                #   then the penalty is 0.75
```

Comment:
I think these should read `the boost is 0.5`

Reply:
done